File: web/src/lib/encryption.ts
import { webcrypto } from 'crypto';

const ALGORITHM = 'AES-GCM';
const IV_LENGTH = 12;
const TAG_LENGTH = 128;

function getEncryptionKey(): string {
  const key = process.env.ENCRYPTION_KEY;
  if (!key) throw new Error('ENCRYPTION_KEY environment variable is not set');
  return key;
}

async function deriveKey(secret: string): Promise<CryptoKey> {
  const encoder = new TextEncoder();
  const keyMaterial = await webcrypto.subtle.importKey(
    'raw',
    encoder.encode(secret),
    { name: 'PBKDF2' },
    false,
    ['deriveKey']
  );
  return webcrypto.subtle.deriveKey(
    {
      name: 'PBKDF2',
      salt: encoder.encode('nexus-twitter-bot-salt'),
      iterations: 100000,
      hash: 'SHA-256',
    },
    keyMaterial,
    { name: ALGORITHM, length: 256 },
    false,
    ['encrypt', 'decrypt']
  );
}

export async function encrypt(plaintext: string): Promise<string> {
  const key = await deriveKey(getEncryptionKey());
  const iv = webcrypto.getRandomValues(new Uint8Array(IV_LENGTH));
  const encoder = new TextEncoder();

  const ciphertext = await webcrypto.subtle.encrypt(
    { name: ALGORITHM, iv, tagLength: TAG_LENGTH },
    key,
    encoder.encode(plaintext)
  );

  // Combine IV + ciphertext and base64 encode
  const combined = new Uint8Array(iv.length + new Uint8Array(ciphertext).length);
  combined.set(iv);
  combined.set(new Uint8Array(ciphertext), iv.length);

  return Buffer.from(combined).toString('base64');
}

export async function decrypt(encryptedBase64: string): Promise<string> {
  const key = await deriveKey(getEncryptionKey());
  const combined = Buffer.from(encryptedBase64, 'base64');

  const iv = combined.subarray(0, IV_LENGTH);
  const ciphertext = combined.subarray(IV_LENGTH);

  const decrypted = await webcrypto.subtle.decrypt(
    { name: ALGORITHM, iv, tagLength: TAG_LENGTH },
    key,
    ciphertext
  );

  return new TextDecoder().decode(decrypted);
}

File: encryption.py
"""
Credential encryption/decryption for the Discord Twitter Bot.
Matches the web portal's AES-256-GCM encryption scheme.
"""

import os
import base64
import hashlib
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

IV_LENGTH = 12  # 96 bits for AES-GCM


def _get_key() -> bytes:
    """Derive a 256-bit AES key from the ENCRYPTION_KEY env var using PBKDF2."""
    secret = os.getenv('ENCRYPTION_KEY')
    if not secret:
        raise ValueError("ENCRYPTION_KEY environment variable is not set")

    # Match the web portal's PBKDF2 derivation
    key = hashlib.pbkdf2_hmac(
        'sha256',
        secret.encode('utf-8'),
        b'nexus-twitter-bot-salt',
        100000,
        dklen=32
    )
    return key


def decrypt(encrypted_base64: str) -> str:
    """Decrypt a base64-encoded AES-256-GCM ciphertext."""
    key = _get_key()
    combined = base64.b64decode(encrypted_base64)

    iv = combined[:IV_LENGTH]
    ciphertext = combined[IV_LENGTH:]

    aesgcm = AESGCM(key)
    plaintext = aesgcm.decrypt(iv, ciphertext, None)
    return plaintext.decode('utf-8')


def encrypt(plaintext: str) -> str:
    """Encrypt plaintext using AES-256-GCM. Returns base64-encoded IV+ciphertext."""
    key = _get_key()
    iv = os.urandom(IV_LENGTH)

    aesgcm = AESGCM(key)
    ciphertext = aesgcm.encrypt(iv, plaintext.encode('utf-8'), None)

    combined = iv + ciphertext
    return base64.b64encode(combined).decode('utf-8')
